src/v/iceberg/values_avro.cc

```diff
@@ -0,0 +1,180 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/values_avro.h"
+
+#include "bytes/iobuf_parser.h"
+#include "iceberg/values.h"
+
+#include <avro/Generic.hh>
+#include <avro/GenericDatum.hh>
+#include <avro/Node.hh>
+
+namespace iceberg {
+
+namespace {
+
+// Returns the Avro datum corresponding to the given value. The value is
+// expected to match the given Avro schema.
+//
+// The given value may be null if the schema refers to an optional value.
+avro::GenericDatum
+val_to_avro(const std::optional<value>&, const avro::NodePtr&);
+avro::GenericDatum base_val_to_avro(const value&, const avro::NodePtr&);
+
+// XXX: need to figure out how to correctly instantiate avro::GenericFixed to
+// support the unimplemented fields. Or move to an impl that uses iobufs. Throw
+// for now.
+struct primitive_value_avro_visitor {
+    template<typename ValueT>
+    avro::GenericDatum operator()(const ValueT& v) {
+        return {v.val};
+    }
+    avro::GenericDatum operator()(const string_value& v) {
+        const auto size_bytes = v.val.size_bytes();
+        iobuf_const_parser buf(v.val);
+        return {buf.read_string(size_bytes)};
+    }
+    avro::GenericDatum operator()(const binary_value& v) {
+        std::vector<uint8_t> data;
+        auto bytes = iobuf_to_bytes(v.val);
+        data.resize(bytes.size());
+        for (size_t i = 0; i < bytes.size(); i++) {
+            data[i] = bytes[i];
+        }
```

Comment:
how big do you think this will be? you can consume_from with the iobuf input iterator directly from the fragments in the input v.val iobuf into the vector.

Reply:
In general, for general Avro data, this is unbounded. Though note that in the short term we're only going to be using this code for manifests, which I expect to be small.

The real long term answer is that we probably need an Avro parser in the same way we have one for protobuf. Avro C++ uses contiguous containers and contiguous strings.

Good to know about the input iterator.